src/sentry/tasks/deletion/hybrid_cloud.py

```diff
@@ -0,0 +1,182 @@
+from hashlib import sha1
+from typing import Any, List, Tuple, Type
+from uuid import uuid4
+
+from django.apps import apps
+from django.db.models import Manager, Max
+
+from sentry.db.models.fields.hybrid_cloud_foreign_key import (
+    HybridCloudForeignKey,
+    HybridCloudForeignKeyCascadeBehavior,
+)
+from sentry.models import TombstoneBase
+from sentry.silo import SiloMode
+from sentry.tasks.base import instrumented_task
+from sentry.utils import json, redis
+
+
+def deletion_silo_modes() -> List[SiloMode]:
+    cur = SiloMode.get_current_mode()
+    result: List[SiloMode] = []
+    if cur != SiloMode.REGION:
+        result.append(SiloMode.CONTROL)
+    if cur != SiloMode.CONTROL:
+        result.append(SiloMode.REGION)
+    return result
+
+
+def get_watermark_key(prefix: str, field: HybridCloudForeignKey) -> str:
+    return f"{prefix}.{field.model._meta.db_table}.{field.name}"
+
+
+def get_watermark(prefix: str, field: HybridCloudForeignKey) -> Tuple[int, str]:
+    with redis.clusters.get("default").get_local_client_for_key("deletions.watermark") as client:
+        key = get_watermark_key(prefix, field)
+        v = client.get(key)
+        if v is None:
+            result = (0, uuid4().hex)
+            client.set(key, json.dumps(result))
+            return result
+        return tuple(json.loads(v))
+
+
+def set_watermark(
+    prefix: str, field: HybridCloudForeignKey, value: int, prev_transaction_id: str
+) -> None:
+    with redis.clusters.get("default").get_local_client_for_key("deletions.watermark") as client:
+        client.set(
+            get_watermark_key(prefix, field),
+            json.dumps((value, sha1(prev_transaction_id.encode("utf8")).hexdigest())),
+        )
+
+
+def chunk_watermark_batch(
+    prefix: str, field: HybridCloudForeignKey, manager: Manager, *, batch_size: int
+) -> Tuple[int, int, bool, str]:
+    lower, transaction_id = get_watermark(prefix, field)
+    upper = manager.aggregate(Max("id"))["id__max"] or 0
+    batch_upper = min(upper, lower + batch_size)
+    return lower, upper, batch_upper + batch_size < upper, transaction_id
```

Comment:
Is the third offset here indicating if there is more work to be done still? If so should the condition be `batch_upper < upper`?

Reply:
Good question!  It's, more like, "should deletes continue to process now or await a future batch?"

The idea is to avoid overly aggressively processing "trickles".  If there are, say, 3 deletes every second, this job could recursively process continuously, but it would be doing tons of small batches.  This ensures that recursive, immediate processing stops when there isn't at least an entire batch of work, in which case we await the next schedule batch.